### DISCUSSION TITLE
Multiple Callbooks lookup

### DISCUSSION THREAD

**Discussion by R1BLH** [wavelog/wavelog#494] [Ideas]
Currently, Wavelog only supports lookups in a single callbook, which is configured in the configuration file.
However, there is a problem that not all callsigns are available in all callbooks. Also, some callbooks have a limit on the number of queries per day if there is no paid subscription (on qrz.com -- 100 queries per day).

It would be nice if the ability to work with multiple callbooks could be implemented using the following algorithm:
1. access to the callbooks is set in the configuration file (login/password/etc.)
2. the order of callbook lookups is configured
3. during the automatic lookup, the callbook is requested according to the queue configured in point 2. If the callsign data is not received from the current callbook, the next callbook is requested until the data is received or the queue runs out.

In addition, this algorithm allows working with national callbooks by receiving data from them first.

**Comment by HB9HIL** [comment]
I like the idea. Currently nothing like that is planned but we are open for suggestions  :D

**Reply by R1BLH** [reply]
This idea also ties in with support for national callbooks like qrz.ru that we talked about yesterday (and others that have APIs).
It would be great if wavelog could document what parameters to pass when requesting data from a callbook, such as qrz.com, and how to pass them. This would help to write modules for national callbooks.

**Reply by AndreasK79** [reply]
@R1BLH I rewrote a lot of the callbook stuff a few months ago. While doing that, I added qrzcq as a callbook, and I did look into qrz.ru. I have a library for that on my PC, but it is not done yet. I can look into what is missing, and finish it up.

**Reply by R1BLH** [reply]
@AndreasK79 That is really good news! Thank you very much!
Do you have any idea how to combine several callbooks into one request (in the queue one after the other, until information about the callsign is received)? This would be useful because callsign information is not present in all callbooks. For example, hamqth has it, but qrz.com does not, as well as other variants. It happens that some callsigns are only available in national callbooks.

**Reply by AndreasK79** [reply]
@R1BLH it should not be too hard I think. Creating a loop that stops if a result is found.

**Reply by AndreasK79** [reply]
@R1BLH I see you need a .ru email address for qrz.ru, so I can't test this implementation at all.
What I also see, is that very little information is returned from the API:
https://qrz.ru/help/api/xml

I can make a pull request with support for QRZ.ru, but I have no way of testing it.

**Reply by R1BLH** [reply]
> I can make a pull request with support for QRZ.ru, but I have no way of testing it.

Strange. 
But the solution to the access problem can be quite simple.
1. I have login/password for access to XML API qrz.ru, I can give it to you. This is probably the easiest way.
2. You can get an account in qrz.ru, you can get an email in .ru zone on any free service (yandex.ru/mail.ru/...). But this way is more complicated than point 1.
3. About the data that can be obtained from XML API qrz.ru: photo,
Name, QTH, Gridsquare, RDA (CNTY), Mail address, QSL info.
For example, here are screenshots from EWLog Mobile program (I use it for offline, then import ADIF into Wavelog).

![photo_2025-02-02_23-55-04 (2)](https://github.com/user-attachments/assets/81228301-2ecb-44e9-b7c1-9fd7dda683b4)
![photo_2025-02-02_23-55-04](https://github.com/user-attachments/assets/0c162b69-f203-4506-93ed-ba027be7dddf)


P.S. it looks like information at the XML API help page is very old.

**Reply by R1BLH** [reply]
> it should not be too hard I think. Creating a loop that stops if a result is found.

It would be a really cool feature. By default now I'm using hamqth (hamqth provides the gridsquare), but often callsign is missing in the hamqth and present in the qrz,com.

**Reply by R1BLH** [reply]
> What I also see, is that very little information is returned from the API:
> https://qrz.ru/help/api/xml



Here is the result of the request for data about my callsign (R1BLH):

![изображение](https://github.com/user-attachments/assets/2c05f35e-bd24-4b34-9600-ee83c0dd324d)

You are right, there is not much useful information, but the main fields in the QSO record could be filled.

`    <call>R1BLH</call>
`
surname and name in russian
```
    <surname>Скольский</surname>
    <name>Михаил</name>
```

surname and name in english
```
    <esurname>Skolsky</esurname>
    <ename>Michael</ename>
```

postal address
```
    <zip>197375</zip>
    <city>Санкт-Петербург</city> -- also name of QTH
    <street>a/я 129</street>
```

gridsquare
 `   <qthloc>KP50da</qthloc>`
ITU and CQ zones
```
    <itu_zone>29</itu_zone>
    <cq_zone>16</cq_zone>
```

RDA (CNTY)
`    <state>SP-12</state>`
also code of the primary administrative division (STATE) could be filled (this is first part of the RDA code -- SP).


Full text of the requrst result is below:

```
<?xml version="1.0"?>
<QRZDatabase version="1.0" xmlns="http://api.qrz.ru/namespace">
<Callsign>
    <call>R1BLH</call>
    <type>0</type>
    <usage_type>0</usage_type>
    <othercall></othercall>
    <country_id>1</country_id>
    <country>Россия</country>
    <qslvia><![CDATA[bureau or direct]]></qslvia>
    <surname>Скольский</surname>
    <name>Михаил</name>
    <esurname>Skolsky</esurname>
    <ename>Michael</ename>
    <birthday>03.09.1971</birthday>
    <zip>197375</zip>
    <city>Санкт-Петербург</city>
    <street>a/я 129</street>
    <state>SP-12</state>
    <region_id>92</region_id>
    <url>https://cats-shadow.cats-home.net</url>
    <icq>6625755</icq>
    <skype>cats_shadow</skype>
    <class>2</class>
    <qthloc>KP50da</qthloc>
    <latitude>60</latitude>
    <longitude>30.25</longitude>
    <qth_type>qth</qth_type>
    <itu_zone>29</itu_zone>
    <cq_zone>16</cq_zone>
    <is_eqsl>N</is_eqsl>
    <is_lotw>Y</is_lotw>
    <is_mailqsl>Y</is_mailqsl>
    <created>02.06.2022 13:03:48</created>
    <lastedit>23.12.2024 17:39:56</lastedit>
    <lookup>2789</lookup>
    <biography><![CDATA[<p style="text-align: center;"></p><p>
	<strong>Немного о себе.</strong></p><p>
	Основной QTH — Санкт-Петербург.</p><p>
	Начинал заниматься радио ещё в школе. В 1986..1988 был оператором на коллективках UO4OXA (школьной), UO4OYK (городского клуба ДОСААФ), тогда же был наблюдательский позывной 
	<strong>UO5-039-517</strong>.<br>
	Вот только сейчас получилось добыть немного времени и получить позывной (1 июля 2022г).</p><p>
	Один из моих интересов — распределённые некоммерческие социальные сети (
	<a href="https://fediverse.observer/">Fediverse</a>), предпочитаю использовать их вместо коммерческих (ВК/FB/...).
	<br>
	Поднял 
	<a href="https://friends.radio-spb.site/" target="_blank">русскоязычный узел Федиверс для радиолюбителей</a> (на движке <a href="https://friendi.ca/" target="_blank">Friendica</a>). Мой аккаунт там <a href="https://friends.radio-spb.site/profile/r1blh" target="_blank">@r1blh@friends.radio-spb.site</a>. Там же читаю подписки с <a href="https://mastodon.radio" target="_blank">радиолюбительского узла Mastodon</a>. <br>
	Регистрация модерируемая (рекомендуется идентифицировать себя по емэйл/телеграм/матрице — сообщить позывной).</p><p>
	Pixelfed: 
	<a href="https://pixelfed.cats-home.net/cats-shadow" target="_blank">@cats-shadow@pixelfed.cats-home.net</a>  (на своём собственном узле <a href="https://pixelfed.org/" target="_blank">Pixelfed</a>, аналог Instagram).</p><p>
	matrix: @cats-shadow:cats-home.net
	<br>
	email: r1blh@yandex.ru
	<br>
telegram: @cats_shadow</p><p>
	В качестве лога использую 
	<a href="https://www.wavelog.org/" target="_blank">Wavelog</a> (форк <a href="https://www.magicbug.co.uk/cloudlog/" target="_blank">Cloudlog</a>, которым пользовался ранее), занимаюсь его русской локализацией.</p><p style="text-align: center;">
	<a href="https://wavelog.cats-home.net/index.php/visitor/R1BLH"><img src="https://wavelog.cats-home.net/index.php/staticmap/render/R1BLH" alt="Публичная страница лога R1BLH (Wavelog)" style="width: 469px;"></a></p><p>
	Мысли свои про логи и системы электронных QSL свёл заметку, буде кому любопытно — 
	<a href="https://telegra.ph/Apparatnye-zhurnaly-lyubitelskih-radiostancij-01-19">вот тут.</a> Периодически обновляю.</p><p>
	<strong>
	Про QSL.
	</strong></p><iframe height="270" width="600" frameborder="0" src="https://wavelog.cats-home.net/index.php/widgets/oqrs/R1BLH?theme=default&slug=R1BLH">
</iframe><p>	Со школьных времён бумажные QSL-карточки вызывают тёплые чувства, потому всегда рад получить и отправить их. Но загвоздка в том, что раньше всё ходило через бюро (по крайней мере тогда, в Молдавии), а сейчас как QSL-почта ходит, скажем прямо, как получится. Да и электронные варианты прогресс принёс.</p><p>
	Так что с QSL история такая:</p><ul>
	
<li>Отвечаю на все полученные QSL тем методом, которым получена карточка, без каких-либо условий.</li>	
<li>Бумажные QSL-карточки первым отправляю сам:
	
<ul>
		
<li>через бюро, если могу для себя подтвердить, что адресат пользуется QSL-бюро</li>		
<li>почтой:
		
<ul>
			
<li>если мне очень нужно подтверждение, но ни электронного, ни через бюро вариантов нет</li>			
<li>если мне хочется QSL-карточку в коллекцию</li>		</ul></li>	</ul></li>	
<li>QSL связей в цифровых режимах по умолчанию отправляю электронные, но на бумажные отвечаю. Также можно запросить QSL карточку через OQRS (см. ниже).</li></ul><p>
	<strong>
	Настроен OQRS для запроса QSL-карточек</strong></p><ul>
	
<li>на <a href="https://wavelog.cats-home.net/index.php/oqrs" target="_blank">домашнем Wavelog</a> (см. выше) </li>	
<li>на <a href="https://clublog.org/logsearch/R1BLH" target="_blank">ClubLog</a></li></ul><p>
	Через бюро запрошенные QSL отправлю без условий, директом — за стоимость конверта с марками. :)</p><p>
	<strong>
	Касательно электронных систем QSL:</strong></p><ul>
	
<li>LoTW — автоматическая синхронизация</li>	
<li>clublog.org — автоматическая загрузка</li>	
<li>hrdlog.net (ham365.net) — автоматическая загрузка</li>	
<li>qso.su — лог заливается вручную</li>	
<li>logradio.ru — лог заливается вручную</li>	
<li>hamlog.online после недавних событий как-то потерял осмысленность, превратившись в ещё одну "вещь в себе". Да и огорожен он со всех сторон, все манипуляции только вручную. Мирился с этим, считая его осмысленной локальной системой подтверждения QSO, но всё рано или поздно заканчивается. Как только появится альтернатива, подумаю о переводе заливки в hamlog.online на "остаточный принцип/по запросу/мне интересен опубликованный там диплом".</li>	
<li>qrz.com — периодически синхронизируется с LoTW</li>	
<li>отказался от eQSL.cc по ряду причин, хоть этот сервис и поддерживается Wavelog "из коробки"</li></ul><hr><p>
	И немного фото. :)</p><p>
	Рисунок для QSL карточки авторства дочери. Говорит, что кота рисовала с меня. :)</p><p>
	<img src="https://static.qrz.su/callbook/bc86/1d766b8486cbacd5cf508b5a2e32f114.png" alt=""></p><p>
	Кошь по имени Маркиза принимает деятельное участие во всём происходящем.</p><p>
	<img src="https://static.qrz.su/callbook/bc86/0092024674ee3d44e142a27936c9d6c2.jpg" alt="" style="width: 244px;"> <img src="https://static.qrz.su/callbook/bc86/9aa9138aef7fcc65700e6393f370522d.jpg" alt="" style="width: 409px;"></p><p>
	Шек на фазенде</p><p>
	<img src="https://static.qrz.su/callbook/bc86/7cda0201bd67ccc7a10090c59b4987e1.jpg" alt="" style="width: 412px;"></p><p>
	Или в лесу</p><p>
	<img src="https://static.qrz.su/callbook/bc86/b010d6140e2db8b513e1c196a9c5f234.jpg" alt="" style="width: 249px;"></p><p>
	73! Michael
	<br>
	R1BLH</p>]]></biography>
</Callsign>
<Files>
    <file>https://static.qrz.su/callbook/bc86/1d766b8486cbacd5cf508b5a2e32f114.png</file>
    <file>https://static.qrz.su/callbook/bc86/4de6886ce186b3a1c8613a01d652fc31.jpg</file>
    <file>https://static.qrz.su/callbook/bc86/0092024674ee3d44e142a27936c9d6c2.jpg</file>
    <file>https://static.qrz.su/callbook/bc86/9aa9138aef7fcc65700e6393f370522d.jpg</file>
    <file>https://static.qrz.su/callbook/bc86/7cda0201bd67ccc7a10090c59b4987e1.jpg</file>
    <file>https://static.qrz.su/callbook/bc86/b010d6140e2db8b513e1c196a9c5f234.jpg</file>
</Files>
<Session>
    <session_id>.................................</session_id>
    <GMTime>Sun, 02 Feb 2025 21:33:48 +0000</GMTime>
</Session>
</QRZDatabase>
```

**Comment by AndreasK79** [comment]
@R1BLH thanks for the information. It does seem that their api description is outdated. I will update with the information you gave me, and when it is ready for testing, I'll contact you,

**Reply by R1BLH** [reply]
@AndreasK79 That's great!
If anything -- I'm willing to help with both access and testing.

**Reply by AndreasK79** [reply]
@R1BLH ready for testing: https://github.com/wavelog/wavelog/pull/1610

**Reply by R1BLH** [reply]
> @R1BLH ready for testing: #1610
I've commented there

**Comment by vastsea-wuji** [comment]
I thought it was a great idea as I am not a QRZ subscription member 😄

**Reply by AndreasK79** [reply]
Well, we have hamqth and qrzcq as well. QRZ.ru is untested, so I don't know if it works.

**Comment by AndreasK79** [comment]
@R1BLH now that qrz.ru support is done for now, I can look into doing several callbooks. One issue that was mentioned to me, is that it might take a long time doing normal qso logging when you want to query several callbooks.

If you really want this, I can take a look, however, that will be for another day.

**Reply by R1BLH** [reply]
> If you really want this, I can take a look, however, that will be for another day.

Thanks for what you've done!
Yes, such functionality would be handy, and the best option is to set the polling order and stop when the data is received.

By the way, I noticed an interesting situation. I have configured a query from qrz.ru. And it fills in the data for callsigns from the USA, which I have never worked with before. It does not fill in the data for callsigns from other countries. I wonder where this data comes from?

**Reply by R1BLH** [reply]
@AndreasK79 
Hi!
I was thinking about my suggestion about cyclically querying data from multiple callbooks. It's still not a good idea.
Maybe a better option would be not only to follow a link to a profile in a callbook as it is now, but also to have similar buttons to fill in fields in the QSO entry form from the corresponding callbook.
For example, the default setting is to request data from hamqth.com.
During a QSO you can manually request data from the desired callbook (qrz.com etc.) and overwrite the data in the QSO entry form.
How difficult is it to implement such a feature?

**Comment by vastsea-wuji** [comment]
QRZ's XML queries seem to be unlimited, except that non-subscribers return fewer fields. It only limits the number of page searches
